### ISSUE TITLE
Undefined variable references not always checked in rules

### ISSUE THREAD

**Issue by mrrodriguez** [rbrush/clara-rules#122]
I have been bitten by an issue that can sneak up where the Clara compiler does not catch undefined variable bindings being used.  I've been able to minimize this issue to a case of variable binding validity not being checked on `:join-filter-expressions` constraints.  I'm not completely sure this is the full extent of the issue though.

To give some motivation via an example, the following query is "successfully" compiled by Clara, but the variable binding `?bogus` has not been defined and will always be nil.
```clj
(let [q (dsl/parse-query []
                         [[?ts <- (acc/all) :from [Temperature (and ?bogus (< ?bogus temperature))]]])]
  (-> (mk-session [q] :cache false)
      (insert (->Temperature 120 "MCI"))
      fire-rules
      (query q)))

;= ({:?ts []})
```

I have tracked down where the Clara compiler attempts to check for invalid variable binding references, which is in the `clara.rules.compiler/check-unbound-variables` function.  However, this function only checks `:test` type expressions because they are simple to check.  They do not introduce any new variable bindings, so all variable bindings referenced within them must previously be bound.

I think this check is a little too simplistic due to cases like my example above.  The 
`:join-filter-expressions` constraints are never checked in this case to ensure the variable bindings referenced are defined.  If the expressions are not sensitive to nil, as I demonstrated above, then they  will always evaluate the variable as nil.  This can make functionally incorrect behavior that is hard to detect.

I was going to attempt to provide a patch for this, but I thought I'd seek feedback instead (first perhaps).  The way the `c.r.c/check-unbound-variables` function currently behaves, it is unaware of the context of a variable binding reference.  If it were made to be aware of the context, it may be able to detect variable binding references that are not in an equality, aka unifying, expression and catch cases like above.  This may work if unifications is strictly limited to expressions of the form 
(note: `=` vs `==` vs qualified variants isn't relevant here):
```clj
(= ?v blah) ;; or
(= blah ?v)
```
I think this is the general assumption made by the compiler.  If the function was enhanced to be aware of variable binding context like this, it could be widened to check conditions of types other than only `:test`.
With this change, unbound variable references like my example above should be able to be caught.

**Comment by rbrush** [comment]
Agreed that we should expand or refactor check-unbound-variables to catch this scenario. That particular function was added to handle invalid test expression cases, but we should handle this for all expressions. In other words, check-unbound-variables is basically incomplete in its current form.

If you were looking at this and interested in taking a pass, that would be welcome. Otherwise I'll go ahead and work on this myself since anything that can lead to unexpected behavior without producing an error is pretty bad.

**Comment by mrrodriguez** [comment]
I can try to take a look at this tonight and see if I can come up with a good solution.  I'll post back here with my results - hopefully in the form of a working PR.

**Comment by mrrodriguez** [comment]
I worked on this some tonight.  I think I have an approach that works, but I will need to clean it up and get some more tests around it to be confident.  I'll have to table that until tomorrow.  
I'm giving you heads up just so you know that this is a WIP and I am hopeful to be able to get a PR by tomorrow.

**Comment by rbrush** [comment]
Cool. Thanks for working on this!

**Comment by rbrush** [comment]
Closing since this was merged and marking for a 0.8.9 release.
